### PR TITLE
TNT-32639 - adding a lightweight version of Turbine along with events…

### PR DESCRIPTION
**Pull request by artur-ciocanu**
Adding a lightweight version of Adobe Launch Turbine engine. The engine is used by Personalization component to render personalized content. 

## Description

The Personalization component uses a lightweight version of Adobe Launch Turbine engine. The engine is used for `rules` executions. The assumption is that our edge will return a list of rules. The rules should contain details like when and how we should render personalized content. Once we managed to render the content we would rely on the DataCollector component to send metadata like Activity, Campaign, Experience ID etc.

## Motivation and Context

Personalization component needs to know when and how to render the personalized content. The Adobe Launch Turbine engine that has been added in this PR has all the necessary ingredients. 

## How Has This Been Tested?

For now it was tested using `sandbox`. The plan is to augment this PR with unit tests and other goodies, once I have green light that the approach I took makes sense.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] I have run the Sandbox successfully.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by artur-ciocanu** [comment]
@Aaronius while I am pleased in how everything works with the lightweight version on Turbine, I am not entirely sure that passing `componentsRegistry` to an `action` is a good idea, so I wanted to ask your opinion.

Overall it's amazing how nicely everything plays together.

CC: @jfkhoury

**Comment by jfkhoury** [comment]
> @Aaronius while I am pleased in how everything works with the lightweight version on Turbine, I am not entirely sure that passing `componentsRegistry` to an `action` is a good idea, so I wanted to ask your opinion.
> 
> Overall it's amazing how nicely everything plays together.
> 
> CC: @jfkhoury

**Comment by Aaronius** [comment]
I've got a bigger review coming, but I agree with @jfkhoury's [comment](https://github.com/adobe/alloy/pull/62#issuecomment-502144944).

File: src/components/Personalization/turbine/index.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+      logger.error(getErrorMessage(event, rule, e.message, e.stack));
+    }
+  };
+
+  buildRuleExecutionOrder(rules).forEach(initEventModule);
+  eventModulesInitialized = true;
+  triggerCallQueue.forEach(triggerCall => triggerCall());
+  triggerCallQueue = null;
+
+  return lastPromiseInQueue;
```
Comment (Aaronius) [inline src/components/Personalization/turbine/index.js]: You should be able to remove `lastPromiseInQueue`.

**Comment by Aaronius** [comment]
@artur-ciocanu, with the module provider stuff, I think there's a decent amount of this code that's unnecessary in Alloy (and, actually, we'll be removing in Turbine once we move to a build system that uses a proper JS bundler). I'll create a new branch and hammer away a little bit and submit a PR based on this one.

File: src/components/Personalization/turbine/index.js (hunk 1 above)
Comment (artur-ciocanu) [reply src/components/Personalization/turbine/index.js]: Done.

File: src/components/Personalization/flicker/index.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+const STYLE_TAG = "style";
+const CLASS_ATTRIBUTE = "class";
+
+const buildClassName = prehidingSelector => {
+  return `${PREFIX}-${hash(prehidingSelector)}`;
+};
+
+export const prehideSelector = prehidingSelector => {
+  const className = buildClassName(prehidingSelector);
+  const node = createNode(STYLE_TAG, { [CLASS_ATTRIBUTE]: className });
```
[the remark this answers is not in the document]
Comment (artur-ciocanu) [reply src/components/Personalization/flicker/index.js]: @Aaronius I have added a in memory cache from `prehidingSelector` to `style` tag, so we could get rid of those once the `awaitSelector` times out or the content has been rendered successfuly.

I have renamed the flicker control functions to `hideElements` and `showElements`.

**Comment by artur-ciocanu** [comment]
@jfkhoury and @Aaronius could you please take another look. Thank you!

File: src/components/Personalization/actions/setHtml.js [hunk 3]
```diff
@@ -10,20 +10,20 @@ OF ANY KIND, either express or implied. See the License for the specific languag
...
 export default collect => {
   return (settings, event) => {
     const { element, prehidingSelector } = event;
-    const { content, metadata } = settings;
+    const { content, meta } = settings;
 
     // this is a very naive approach, we will expand later
     element.innerHTML = content;
 
     // after rendering we should show remove the flicker control CSS classes
```
Comment (Aaronius) [inline src/components/Personalization/actions/setHtml.js]: Typo: "show remove"

Also, is the "CSS classes" part accurate/applicable still?

**Comment by artur-ciocanu** [comment]
@Aaronius I have adjusted the code comments and I have added a few unit tests for events and actions.

File: src/components/Personalization/actions/setHtml.js (hunk 3 above)
Comment (artur-ciocanu) [reply src/components/Personalization/actions/setHtml.js]: @Aaronius the idea was to have something like `hideElements()` that match a particular selector and the counterpart is `showElements()` that match a selector.

If you think this is confusing, I'll try to come up with different names.

**Pull request merged**